pkg/gcpwatch/tracker.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcpwatch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/iam/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+type DependencyTracker struct {
+	fetcher Fetcher
+
+	mutex        sync.Mutex
+	gcpResources map[gcpResouceKey]*dependenciesByResource
+
+	controllers map[string]*ControllerRegistration
+}
+
+type ControllerRegistration struct {
+	tracker       *DependencyTracker
+	controllerKey string
+	queue         chan event.GenericEvent
+}
+
+func (t *DependencyTracker) RegisterController(key string, queue chan event.GenericEvent) *ControllerRegistration {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if _, exists := t.controllers[key]; exists {
+		klog.Fatalf("multiple controllers with key %q", key)
+	}
+
+	registration := &ControllerRegistration{
+		controllerKey: key,
+		queue:         queue,
+		tracker:       t,
+	}
+
+	t.controllers[key] = registration
+
+	return registration
+}
+
+type gcpResouceKey struct {
+	Kind     string
+	External string
+}
+
+type dependenciesByResource struct {
+	mutex        sync.Mutex
```

Comment:
Ditto

Reply:
Also if there is *any* chance we copy this object then this should be a pointer.

Reply:
I think if we always use a pointer receiver for the structs (which we are doing), we're fine against the copy lock. In my experience too, `go vet` will yell at you if you try to copy a mutex otherwise